### PR TITLE
provided a alt list of states for filtering Cylc 8 tasks

### PR DESCRIPTION
**Pull request by wxtim**
These changes partially address #4782 

**Features check-list**
In `cylc review`...
- [x] Provide a different set of task state filters for Cylc 8 and Cylc 7.
- [x] Turn on waiting tasks by default if workflow is Cylc 8 (still under debate at #4782 )
- [x] Don't show tasks without jobs

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Does not need tests - to be manually tested (see below)
- [x] Appropriate change log entry included.
- [x] No documentation update required.

## How I tested 
```ini
# global.cylc (end of)
[platforms]
    [[submitfail]]
        hosts = notahost
    [[submitted]]
        hosts = exvcylcdev01
        install target = exvcylcdev01
        communication method = poll
        submission polling intervals = P1Y
```

```ini
# flow.cylc
[scheduler]
    allow implicit tasks = true
[scheduling]
    initial cycle point = 1
    [[dependencies]]
        R1 = """
             expired & preparing & submitted  => waiting
             running & submit-failed & succeeded & failed =>waiting
            """

[runtime]
    [[submitted]]
        platform = submitted
    [[submit-failed]]
        platform = submitfail
    [[running]]
        script = while true; do sleep 30; done
    [[failed]]
        script = exit 1
```

and add the following hack to Cylc flow:
```diff
diff --git a/cylc/flow/task_job_mgr.py b/cylc/flow/task_job_mgr.py
index ab4174b55..9f3eeadba 100644
--- a/cylc/flow/task_job_mgr.py
+++ b/cylc/flow/task_job_mgr.py
@@ -1082,6 +1082,11 @@ class TaskJobManager:
         Return itask on a good preparation.
 
         """
+        if 'preparing' in itask.identity:
+            LOG.critical('Prevent task called preparing ever being marked as prepped.')
+            return False
+
+
         if itask.local_job_file_path:
             return itask

```

Note: Don't forget to check that the change hasn't done anything bad to Cylc Review's ability to read Cylc flow.

Another Note: Recent versions of the wrapper script may re-direct you to the wrong version of `cylc review`, I found it worthwhile running explicitly from a path - i.e. `${HOME}/cylc-7/bin/cylc review start`

### CODE REVIEW

File: lib/cylc/review.py [hunk 1]
```diff
@@ -51,6 +51,18 @@
...
 
+CYLC8_TASK_STATUSES_ORDERED = [
+    'expired',
+    'failed',
+    'preparing',
+    'running',
+    'submitted',
+    'submit-failed',
+    'succeeded',
+    'waiting',
```
Comment (datamel) [inline lib/cylc/review.py:62]: Could use string literals from task states (as the cylc7 states do) but as legacy probably not very important.

File: lib/cylc/review.py [hunk 2]
```diff
@@ -302,18 +314,27 @@ def taskjobs(
         else:
             page = 1
 
+        # Set list of task states depending on Cylc version 7 or 8
+        from sqlite3 import ProgrammingError
```
Comment (datamel) [inline lib/cylc/review.py:318]: Did this import need to be moved to the top?

Comment (datamel) [inline lib/cylc/review.py:318]: Did this import need to be moved to the top?

File: lib/cylc/review.py [hunk 3]
```diff
@@ -302,18 +314,27 @@ def taskjobs(
...
+                task_statuses_ordered = TASK_STATUSES_ORDERED
+        except ProgrammingError:
+            task_statuses_ordered = TASK_STATUSES_ORDERED
+
         # get selected task states
         if not task_status:
             # default task statuses - if updating please also change the
             # $("#reset_task_statuses").click function in cylc-review.js
-            task_status = [state for state in TASK_STATUSES_ORDERED
-                           if state != TASK_STATUS_WAITING]
```
Comment (datamel) [inline lib/cylc/review.py:310]: This `TASK_STATUS_WAITING` can be removed from the import at the top now.

Comment (datamel) [inline lib/cylc/review.py:310]: This `TASK_STATUS_WAITING` can be removed from the import at the top now.

File: lib/cylc/review.py [hunk 4]
```diff
@@ -302,18 +314,27 @@ def taskjobs(
...
+        # Set list of task states depending on Cylc version 7 or 8
+        from sqlite3 import ProgrammingError
+        try:
+            if self.suite_dao.is_cylc8(user, suite):
+                task_statuses_ordered = CYLC8_TASK_STATUSES_ORDERED
+            else:
+                task_statuses_ordered = TASK_STATUSES_ORDERED
+        except ProgrammingError:
+            task_statuses_ordered = TASK_STATUSES_ORDERED
+
```
Comment (datamel) [inline lib/cylc/review.py:326]: Minor suggestion to replace two assignments with one.
```suggestion
        task_statuses_ordered = TASK_STATUSES_ORDERED
        try:
            if self.suite_dao.is_cylc8(user, suite):
                task_statuses_ordered = CYLC8_TASK_STATUSES_ORDERED
        except ProgrammingError:
            pass
```

**Review by datamel — COMMENTED** [review]
Minor points, working as expected for me. Thanks @wxtim.

File: lib/cylc/review.py (hunk 1 above)
Comment (wxtim) [reply lib/cylc/review.py:62]: but it would have to get these from Cylc 8?

Comment (datamel) [reply lib/cylc/review.py:63]: Sorry, I regretted this comment and should have deleted it, 'twas a fleeting thought. I meant declaring them as string literals as the cylc7 ones are in `task_state.py`. I think this is overkill and not a good use of time.

**Review by datamel — APPROVED** [review]

**Pull request merged**